crates/rust-analyzer/src/handlers.rs

```diff
@@ -1910,3 +1932,34 @@ fn run_rustfmt(
         Ok(Some(to_proto::text_edit_vec(&line_index, diff(&file, &new_text))))
     }
 }
+
+/// Searches for the directory of a Rust crate with a given name in the directory tree
+/// of the root file of that crate.
+///
+/// # Arguments
+///
+/// * `crate_name`: The name of the crate to search for. This should be a `Some` value if
+///   a crate name has been specified, or `None` if no crate name has been specified.
+/// * `root_file_path`: The path to the root file of the crate.
+///
+/// # Returns
+///
+/// An `Option` value representing the path to the directory of the crate with the given
+/// name, if such a crate is found. If no crate with the given name is found, this function
+/// returns `None`.
+fn crate_path(crate_name: Option<&String>, root_file_path: VfsPath) -> Option<VfsPath> {
+    crate_name.and_then(|crate_name| {
+        let mut root_path = root_file_path;
+        while let Some(path) = root_path.parent() {
+            if let Some((name, _)) = path.name_and_extension() {
+                if name.starts_with(crate_name.as_str()) {
+                    return Some(path);
+                }
+            } else {
+                break;
+            }
+            root_path = path;
+        }
+        None
+    })
+}
```

Comment:
Okay, so this will not work too well with crates that are renamed in the cargo toml causing their folder name to differ from the crate name. I think a better approach here is to assume the classic library structure, go up two steps and check whether the `Cargo.toml` is where we expect it. That is check that `{parentPath}/../Cargo.toml` exists, then return `{parentPath}/../..`, so we assume `../crate/src/lib.rs` as the crate root (which is almost always the case for libraries)

Reply:
nice, I changed the code meet your suggestion, it made the code simpler 👍